In [1]:
# Import the modules
import numpy as np
import pandas as pd
from pathlib import Path
from sklearn.metrics import confusion_matrix, classification_report

---

## Split the Data into Training and Testing Sets

### Step 1: Read the `lending_data.csv` data from the `Resources` folder into a Pandas DataFrame.

In [2]:
# Importing pandas library
import pandas as pd

# Reading the CSV file from the Resources folder into a Pandas DataFrame
df = pd.read_csv('Resources/lending_data.csv')

# Reviewing the first few rows of the DataFrame
df.head()


loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  loan_status  
0                 1       22800            0  
1                 0       13600            0  
2                 0       16100            0  
3                 1       22700            0  
4                 1       23000            0

### Step 2: Create the labels set (`y`)  from the “loan_status” column, and then create the features (`X`) DataFrame from the remaining columns.

In [3]:
# Separate the data into labels and features

# Separate the y variable, the labels
y = df['loan_status']

# Separate the X variable, the features
# Drop the 'loan_status' column to keep only the features
X = df.drop('loan_status', axis=1)


In [4]:
# Review the y variable Series

# Display the first 5 entries of the y Series
print(y.head())

# Display the count of unique values in the y Series
print(y.value_counts())

# Check for any missing values in the y Series
print(y.isnull().sum())


0    0
1    0
2    0
3    0
4    0
Name: loan_status, dtype: int64
0    75036
1     2500
Name: loan_status, dtype: int64
0


In [5]:
# Review the X variable DataFrame

# Display the first few rows of the DataFrame to get an overview
print(X.head())

# Display the summary statistics of the DataFrame
print(X.describe())

# Display information about the DataFrame including the data types and non-null values
print(X.info())

# Display the number of missing values in each column
print(X.isnull().sum())


   loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  
0                 1       22800  
1                 0       13600  
2                 0       16100  
3                 1       22700  
4                 1       23000  
          loan_size  interest_rate  borrower_income  debt_to_income  \
count  77536.000000   77536.000000     77536.000000    77536.000000   
mean    9805.562577       7.292333     49221.949804        0.377318   
std     2093.223153       0.889495      8371.635077        0.081519   
min     5000.000000      

### Step 3: Split the data into training and testing datasets by using `train_test_split`.

In [6]:
# Importing the train_test_split module
from sklearn.model_selection import train_test_split

# Split the data into training and testing sets
# Assign a random_state of 1 to the function for reproducibility
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=1)


---

## Create a Logistic Regression Model with the Original Data

###  Step 1: Fit a logistic regression model by using the training data (`X_train` and `y_train`).

In [7]:
# Import the LogisticRegression module from sklearn
from sklearn.linear_model import LogisticRegression

# Instantiate the Logistic Regression model with a random_state of 1
logistic_model = LogisticRegression(random_state=1)

# Fit the model using the training data
logistic_model.fit(X_train, y_train)


LogisticRegression(random_state=1)

### Step 2: Save the predictions on the testing data labels by using the testing feature data (`X_test`) and the fitted model.

In [8]:
# Make predictions using the testing data
y_pred = logistic_model.predict(X_test)


### Step 3: Evaluate the model’s performance by doing the following:

* Generate a confusion matrix.

* Print the classification report.

In [9]:
# Generate a confusion matrix
conf_matrix = confusion_matrix(y_test, y_pred)
print("Confusion Matrix:")
print(conf_matrix)

Confusion Matrix:
[[14926    75]
 [   46   461]]


In [10]:
# Print the classification report
class_report = classification_report(y_test, y_pred)
print("\nClassification Report:")
print(class_report)



Classification Report:
              precision    recall  f1-score   support

           0       1.00      1.00      1.00     15001
           1       0.86      0.91      0.88       507

    accuracy                           0.99     15508
   macro avg       0.93      0.95      0.94     15508
weighted avg       0.99      0.99      0.99     15508



### Step 4: Answer the following question.

**Question:** How well does the logistic regression model predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** 
### Overall Model Performance
The model demonstrates excellent performance in identifying healthy loans (label 0), with both precision and recall being nearly 100%.
For high-risk loans (label 1), the performance is also strong but slightly lower compared to healthy loans. This is indicated by the precision of 86% and recall of 91%.
The overall accuracy of 99% shows that the model is highly effective in classifying the loans correctly.

### Coniderations:
The slight imbalance between the performance on healthy vs. high-risk loans may be due to the imbalanced nature of the dataset (more healthy loans than high-risk ones).
The model might benefit from techniques that handle imbalanced datasets, like oversampling the minority class, undersampling the majority class, or using specific performance metrics like the Area Under the Receiver Operating Characteristic Curve (AUC-ROC).

##Summary
The logistic regression model is performing very well in predicting both healthy and high-risk loans, with particularly strong results for healthy loans and slightly less, but still strong, performance for high-risk loans.


---